### DISCUSSION TITLE
브라우저의 동작과정

### DISCUSSION THREAD

**Discussion by 24tngus** [24tngus/CS_STUDY#20] [CS]
### 📕 웹 브라우저 개념

- 동기(Synchronous)적으로 (HTML+CSS), JavaScript 언어를 해석하여 화면에 보여주는 응용 소프트웨어
- **웹 브라우저**가 **웹 서버**에 **웹 페이지**를 요청하면 서버는 응답하고, 브라우저는 해석한 후 사용자에게 보여줌
- 웹 브라우저 종류 (사용 렌더링 엔진)
  - 네이버 웨일
  - 파이버폭스 (게코(Gecko) 엔진)
  - 사파리 (웹킷(Webkit) 엔진)
  - 크롬 (블링크 엔진)
  
> 📖 참고 📖 크롬 특징
    - 웹킷을 사용하다가 웹킷을 Fork하여 블링크 엔진을 자체적으로 구현해 사용
    - 다른 브라우저와 달리 각 탭마다 별도의 렌더링 엔진 인스턴스 유지 (각 탭마다 독립된 프로세스로 처리)

### 📕 브라우저 구성요소

![](https://velog.velcdn.com/images/24tngus/post/44a2ca90-27f4-4a63-b3a5-772302b1c6e5/image.png)

- (1) 사용자 인터페이스 
  - 사용자 접근 가능 영역
  - 요청한 페이지를 보여주는 창을 제외한 모든 영역
  - 🗒️ 예시
    - URI 입력하는 주소 표시줄
    - 이전/다음 버튼
    - 북마크 메뉴
    - 새로 고침 버튼
    - 홈 버튼 
- (2) 브라우저 엔진
  - 사용자 인터페이스와 렌더링 엔진 사이의 동작 제어
  - Data Storage 참조하여 로컬에 데이터 쓰고 읽는 작업
- (3) 렌더링 엔진
  - 웹 서버로부터 요청하여 응답 받은 HTML 문서를 파싱하여 웹 브라우저 상에 나타냄
  - 🗒️ 예시 : 브라우저가 서버로부터 HTML 문서를 받을 경우
    ➡️ 렌더링 엔진의 HTML 파서와 CSS 파서에 의해 파싱(parsing)됨
    ➡️ DOM 트리와 CSSOM 트리로 변환되고, 렌더 트리로 결합함
    ➡️ 생성된 렌더 트리를 기반으로 브라우저는 웹 페이지를 나타냄
- (4) 통신
  - HTTP 요청 같은, 서버와 통신이 가능하게 하는 네트워크 호출에 사용
- (5) 자바스크립트 해석기 
  - 자바스크립트 코드 해석 후 실행
- (6) UI 백엔드
  - `select`, `input` 등 기본 위젯을 그리는 인터페이스 
- (7) 자료 저장소 
  - Cookie, Local Storage, Indexed DB 등 브라우저 메모리 활용하여 저장 
  
### 📕 [브라우저 동작 과정](https://m.post.naver.com/viewer/postView.nhn?volumeNo=8431285&memberNo=34176766)

![](https://velog.velcdn.com/images/24tngus/post/45473d32-6fd5-45a2-85f5-3a28ca51c811/image.png)

![](https://velog.velcdn.com/images/24tngus/post/2632ef4e-2bd4-4a7b-bd41-fb5532aa9410/image.png)

#### (1) 서버에서 응답 받은 `HTML 데이터` 파싱 ➡️ `DOM Tree` 생성
  - `무엇을` 그릴지 결정
  
> 🗒️ 예시 🗒️ : 브라우저 주소창에 url을 입력할 경우
- 브라우저는 해당 서버에 요청 보냄
- 서버는 응답으로 HTML 문서 전달
- 브라우저가 HTML 문서를 하나씩 파싱하고, 데이터를 화면에 그리는 과정 시작됨
  - 미디어 파일을 만나면, 추가로 요청 보내 받아옴
  - JavaScript 파일 만나면 해당 파일 받아와 실행할 때까지 파싱 멈춤

- 브라우저가 HTML 바이트 데이터를 지정된 인코딩(UTF-8)에 따라 문자열로 변환
- 문자열을 HTML 표준에 따라 토큰(Token)으로 변환
  - <html> ➡️ StartTag: html, </html> ➡️ EndTag: html로 변환
- 토큰을 노드로 바꾸는 과정
  - StartTag:html ➡️ html 노드, EndTag: html 이전 까지 들어오는 토큰을 html 자식 노드로 넣음
  - Tree 모양의 DOM 완성
  
![](https://velog.velcdn.com/images/24tngus/post/d25f410d-d56e-4ae7-a921-a37fea4dc802/image.png)

#### (2) 파싱하는 중 `CSS 파일` 링크 만나면 `CSS 파일` 요청해서 받아서 ➡️ `CSSOM` 생성
  - `어떻게` 그릴지 결정

> 🗒️ 예시 🗒️ 
- HTML 파싱 중 CSS 링크 만나면 CSS 파일 요청 후 받아옴
- HTML 파싱과 유사한 과정으로 Tree 형태의 CSSOM 생성
  - 특성상 자식 노드가 부모 노드의 특성을 계속 이어받는(cascading) 규칙 추가됨 (이외 HTML과 동일)

#### (3) `DOM Tree`와 `CSSOM`이 만들어지면 ➡️`Render Tree` 생성
  - `화면에 그려질 것` 결정

> 🗒️ 예시 🗒️ 
- DOM + CSSOM 합쳐서 RenderTree 생성
  - DOM Tree 중 화면에 실제로 보이는 것만으로 생성
  - CSS에서 display: none의 경우는 포함되지 X
  - HTML의 <head>태그 안의 내용의 경우 포함되지 X
  
![](https://velog.velcdn.com/images/24tngus/post/f795837f-b1d5-4089-abed-3c221bc5f9b1/image.png)

#### (4) `Render Tree`에 있는 각 노드가 화면의 어디에 위치할지 계산하는 Layout 과정 

> 🗒️ 예시 🗒️ 
- Render Tree에 있는 각 노드들이 화면의 어디에 위치할지 계산
  - CSSOM의 스타일 정보로 어떻게 생길지는 알지만, 
  - CSS Box model 사용하여 현재 viewport 기준으로 실제로 놓으려면 어디에 둘지 계산 (position, width, hegiht 등 위치)
  - 

#### (5) 렌더링 트리 그리기
  - `화면에 실제 픽셀을 Paint`

> 🗒️ 예시 🗒️ 
- visibility, outline, background-color 등 눈에 보이는 픽셀 그려짐
- Render Layer이 2개 이상이면, 각 Layer를 paint한 뒤 하나의 이미지로 Composite하는 과정을 추가 후 화면에 그려짐
  
### 📕 [렌더링 엔진 동작 과정](https://d2.naver.com/helloworld/59361)

![](https://velog.velcdn.com/images/24tngus/post/e148dc60-4212-4231-9540-594891f7af4f/image.png)

- (1) HTML 문서 파싱하고, 해당 내부에서 태그를 DOM 노드로 변환함
  - 그 다음 외부 CSS 파일과 함께 스타일 요소도 파싱
- (2) 스타일 정보와 HTML 표시 규칙으로 렌더 트리 형성
  - 색상 또는 면적과 같은 시각적 속성이 있는 사각형 등을 정해진 순서대로 화면에 표시
- (3) 렌더 트리 배치 
  - 각 노드가 화면의 정확한 위치에 표시되는 것
- (4) 렌더 트리 그리기
  - UI 백엔드에서 렌더 트리의 각 노드들로 형상을 만들어 냄
- 빠르게 내용을 표시하기 위해 (1), (2) 기다리지 않고,  (3), (4) 먼저 시작
  
> 📖 참고 📖 웹킷 동작 과정
![](https://velog.velcdn.com/images/24tngus/post/48ec6554-0326-4370-a970-516ed006dea1/image.png)

> 📖 참고 📖 파싱
- 브라우저가 코드를 이해하고 사용할 수 있는 구조로 변환하는 과정
- 문서를 `문맥 자유 문법`에 따라 변환 ➡️ `파싱 트리` or `문법 트리`
  - 문서(소스 코드) ➡️ 어휘 분석(공백, 줄바꿈 등 제거) ➡️ 구문 분석 ➡️ 파싱 트리 ➡️ 변환 ➡️ 기계코드
  ![](https://velog.velcdn.com/images/24tngus/post/081d14ab-8e9a-4572-9c3f-7afa0f74e997/image.png)

> 📖 참고 📖 DOM (문서 객체 모델, Document Object Model)
  - HTML 문서의 객체 표현이고, 외부를 향하는 자바스크립트 같은 HTML요소 연결함
  - [W3C](www.w3.org/DOM/DOMTR)의 명세에 정해져 있음
  - DOM요소 + 속성 노드의 트리 = 파싱 트리 
   - 🗒️ 예시
  ```html
 <html>
  <body>
   <p>Hello World</p>
   <div><img src="example.png" /></div>
  </body>
</html>  
```
![](https://velog.velcdn.com/images/24tngus/post/bc06ac34-945d-4f23-a189-de28100cbc63/image.png)

**Comment by jindream6128** [comment]
확인했습니다.. 프론트 너무 어려운거 같아요....! 
프론트 할때 제일 어려웠던게 Virtual DOM 이었던거 같은데 관련해서도 함께 공부하면 좋지 않을까요?!...

**Reply by 24tngus** [reply]
너무 좋아요 다음에 virtual DOM 자세하게 다뤄봐요

**Comment by berryberries** [comment]
확인했습니다.

**Comment by sujin-park0607** [comment]
확인했습니다!
